### ISSUE TITLE
add property to indicate reoccuring payments to MonetaryAmount (and new type IntervalPeriod)

### ISSUE THREAD

**Issue by Meteor0id** [schemaorg/schemaorg#1999]
Proposing to add a new property "reOccuringPeriod" to "MonetrayAmount", accepting a time period such a, "hourly, daily, weekly, monthly, quarterly, semiAnnually, annually"
Do we already have an type IntervalPeriod for this?

This property will also be of use to PriceSpecification.

**Comment by RichardWallis** [comment]
I do not believe that the concept of [re]occurring payments is an an attribute of a MonetrayAmount,

More a appropriately they are both attributes of some form of payment schedule, for a loan, membership fee etc.

See [LoanOrCredit](https://schema.org/LoanOrCredit) > [loanRepaymentForm](https://pending.schema.org/loanRepaymentForm)  >  [RepaymentSpecification](https://pending.schema.org/RepaymentSpecification)

**Comment by Meteor0id** [comment]
I am specifically keeping in mind the possibility of re-occuring donations (automated payments every moth for example). I recently suggested in another ticket to no longer see DonateAction as a TradeAction but rather see it as a TransferAction (#1998 ).

Re-occuring payments would fit excellent as an attribute to TransferAction in my opinion, and subsecently also to all its more spesific types.

So I agree, it might not fit well on a MonetaryAmount, but let's implement this on TransferAction and its spesific types.

**Comment by RichardWallis** [comment]
I understand your use case. There are a few similar use cases such as regular membership payments, subscription payments, loan repayments.  If we are going to address this it should be with a pattern that fits as many of these as possible.

The pattern around _LoanOrCredit_, I referenced,  is a good start.

I believe the need is to be able to define a _PaymentScheldule_ with properties for payment frequency, amount, plus potentially initial & final payment amounts.    Doing it this way would enable multiple relationships to payment schedules (for example monthly and or weekly membership payment schedules for same thing).

The challenge then would be to identify suitable properties for many types that would benefit from an associated payment schedule.

One though I had, which I am not yet certain of yet, is that _PaymentSchedule_ could be a sup-type of _MonetaryAmount_ - logically not an exact fit, but pragmatically it means the _PaymentSchedule_ could be used wherever a _MonetaryAmount_ is applicable.

**Comment by Meteor0id** [comment]
suggesting the following changes as a start:

- Remove DonateAction as a subtype of TradeAction (subsequently also dropping the properties price and priceSpesification, which is ok since a donation has no price, it has an amount)
- Add DonateAction as a subtype of TransferAction
- Add to TransferAction the property amount, which accepts a type MonetaryAmount (subsequently all subTypes can use this propery to indicate the transfer is monetary)
- Create a new value Thing > Intangible > TimeInterval which expects values such as "hourly, weekly, monthly, quarterly, semi-anually, annualy"
- add to TimeInterval the properties "startDate" and "endDate"
- add to TimeInterval the property "numberOfRecoccences" expecting only a value of type Number
- Add to MonetaryAmount the property "reoccurence" which expects the type as "TimeInterval"


Notice I have kept this seperate from the structure used on loan repayments. After pondering a while, I decided that loan payments have a range of values which would only make sense in such context. If you would like to unify the structures, I suggest bringing it to the evaluation of the pending type LoanOrCredit > loanRepaymentForm > RepaymentSpecification. I would leave it to people who have been working on that type to decide if they can adopt TimeInterval and MonetaryAmount. They are intangables, so they would lend themselfs for it.

Also note I did suggest TimeInterval be used though a property on MonetaryAmount. As @RichardWallis  was saying, it can than be used anywhere when MonetaryAmount is used. He also menstions it logicly does not make perfect sense. But maybe it does if one consideres the total monetary transfer as one whole amount.
Though, the description on the property recoccurence should leave no doubt about one thing; the "amount" specified is the thing which reoccurs, meaning the amount is that of EACH reoccurence, not that to which the recoccerences amount.

**Comment by RichardWallis** [comment]
@Meteor0id 

Three comments on your suggestions:

1. Move _DonateAction_ to be a subtype of _TransferAction_. - I am not in favour of this as I believe donation is a form of altruistic trading, and as described _TransferAction_ is about moving objects from one place to another.  If MonetaryAmount had been in the vocabulary at the time that the Action types were created, I would probably have suggested that _amount_ should have DonateAction in its range, and would support its inclusion now.

2. Add _amount_ as a property of _TransferAction_ - there are two reasons for not doing this:
   -   It would give the false impression that _TransferAction_ is only relevant for money transfers - you can transfer anything, money, land, property, a store (as per the example), etc. (any type defined as a value for the _object_ property.)

   -  Because the _object_ property has an expected type of _Thing_, you can assign any type to the value of that property, including _MonetaryAmount_ which would satisfy your use case.

3.  Adding _reoccurrence_ as a property to _MonetaryAmount_, expecting _TimeInterval_ as a value.  To my mind, the logic here is the wrong way around.  An amount of money is an attribute of a single, or multiple payment schedule; not that an amount of money repeats itself in some pattern.    From a payment point of view, a simple repeat time period is too simplistic.  Often, enough to need to be considered as part of the use case, payment schedules have different starting, repeating, and end payment amounts.

Reconsidering my earlier comments, if we did create a _PaymentSchedule_ type, it should not be a subtype of _MonetaryAmount_.  It should be a type in its own right, following the pattern established for loans, that would be included in the range of some relevant properties that currently also expect a MonetaryAmount.

**Comment by Meteor0id** [comment]
1. My issue with DonateAction is that it takes a price property rather than an amount property, which means having to lie about the intention of the money transfer. Not moving it the DonateAction over but adding the amount property would come a long way, and I would consider that issue solved. 

2. Not sure about this. Money is usually not an object. And amount does not have tobe added to TransferAction, but could be added to some of its more spesific types.

3. Completely agree. As I was writing the previous coment you can tell I was struggeling to explain that the amount shoudl only spesify one payment. I was stuggeling because it made no sense, I had it reversed.

One thing I would like to get your view on still: What about the TimeInterval type? I think that would be useful to many more cases, such as an EventSeries for example. At this time, https://pending.schema.org/loanPaymentFrequency seems very vague, and it might better be replaced by something more general.
For some payments a single TimeInterval would be too simplistic, but RepaymentSpesification currently also has no way of defining an array of values. that is a future concern, if there is a need for such complexities.

I would suggest also allowing a Time value on TimeInterval (makes sense) so that combined with the properties startDate and or endDate, the timeInterval can be deduced to a spesfic point in time when needed, down to the second. (weekly would be exactly 24*7 hours interval, monthly would depent on the day of startDate (or EndDate) and assume every matching daynumber of any month is when the time interval has passed)
A TimeInterval can be interpreted with great precision if used in such manner.

**Comment by Download** [comment]
> weekly would be exactly 24*7 hours interval

Unless, of course, the week was the week in which daylight savings time switched. In countries with DST, one day every year has 23 hours and one day has 25. So basically, daily and weekly also depend on the startDate.
